### PR TITLE
[otbn, sca] Keep shares in distant registers

### PR DESCRIPTION
**Pull request by abdullahvarici**
In FPGA implementation of the earlgrey, blanking is not used for Bignum RF due to high resource consumption and this causes leakage with FPGA experiments. This is an intermediate solution. We are keeping shares in registers with higher Hamming distance between addresses so that we are reducing the probability of reading out both shares when address switches.
    
Signed-off-by: Abdullah Varici <abdullah.varici@lowrisc.org>

### CODE REVIEW

File: sw/otbn/crypto/p256_key_from_seed_sca.s [hunk 1]
```diff
@@ -53,33 +53,35 @@ run_gen_secret_key:
 
   /* Load shares of seed from DMEM.
        [w21,w20] <= dmem[seed0]
-       [w23,w33] <= dmem[seed1] */
+       [w11,w10] <= dmem[seed1] */
   li        x2, 20
   la        x3, seed0
   bn.lid    x2, 0(x3++)
   li        x2, 21
   bn.lid    x2++, 0(x3)
```
Comment (jadephilipoom) [inline sw/otbn/crypto/p256_key_from_seed_sca.s:61]: nit: The `++` here is now unnecessary; same for the similar block below.

**Review by jadephilipoom — APPROVED** [review]
Looks very sensible to me :slightly_smiling_face:

File: sw/otbn/crypto/p256_key_from_seed_sca.s (hunk 1 above)
Comment (abdullahvarici) [reply sw/otbn/crypto/p256_key_from_seed_sca.s]: Thanks! Removed it.

File: sw/device/sca/otbn_vertical/ecc256_keygen_serial.c [hunk 2]
```diff
@@ -135,6 +135,10 @@ void ecc256_en_masks(const uint8_t *enable, size_t enable_len) {
 void ecc256_set_seed(const uint8_t *seed, size_t seed_len) {
   SS_CHECK(seed_len == kEcc256SeedNumBytes);
   memcpy(ecc256_seed, seed, seed_len);
+  // Run PRNG once to synch with capture side. There it is used also to
+  // generate the fixed seed.
+  uint32_t temp[kEcc256SeedNumWords];
+  prng_rand_bytes((unsigned char *)temp, kEcc256SeedNumBytes);
```
[the remark this answers is not in the document]
Comment (abdullahvarici) [reply sw/device/sca/otbn_vertical/ecc256_keygen_serial.c]: Yes, it was already merged as a individual PR.

**Comment by abdullahvarici** [comment]
The changes that I did was breaking other tests like p256_key_from_seed_test. Because of that I copied the routines in p256_key_from_seed_sca.s and made the necessary changes there. Could you review this PR again and merge it if you are approving these changes, @jadephilipoom?

**Comment by jadephilipoom** [comment]
Sorry for the long delay, just got back from vacation :palm_tree: 

Changing to more distant registers is desirable for the core implementation -- I don't think the SCA entrypoint should use different and more hardened routines than the actual implementation, because then the SCA testing will be telling us everything is fine when the real implementation has leakage! I'd prefer to fix the tests to use different registers. Since these routines seem to mostly write results to memory, I hope it shouldn't be too complicated to fix the tests. Would you mind reverting this back to the original version which only changes the registers used in p256.s? I can check it out from there and see if I can quickly fix the failing test.

**Comment by abdullahvarici** [comment]
Thanks @jadephilipoom for your feedback. Actually this issue is only related with the FPGA implementation. So we don't expect to see any leakage in ASIC implementation since we have bignum register file blanking there. That's why I thought it wouldn't be necessary to make this change in all files. But having one entrypoint would be always better, so I reverted the changes.

**Comment by jadephilipoom** [comment]
I fixed the test and also fixed a few of the other call sites for `p256_key_from_seed` (which aren't currently covered by tests, but will be soon). I don't have permission to push to your branch, but if you `git apply` this patch file it should work. (I called it with `.txt` just so github would let me attach it to the comment.)

[use_distant_regs.txt](https://github.com/lowRISC/opentitan/files/11736400/use_distant_regs.txt)

**Comment by vogelpi** [comment]
Thanks for your input @jadephilipoom . Let's try to update the PR after the next milestone / current quiet period to reduce burden on CI. 

@abdullahvarici can you please make sure that other contributors have write access to this branch? We can then update and merge the PR after the quiet period.

**Comment by abdullahvarici** [comment]
Now I've enabled the force-push option to the `use_distant_regs` branch in my repository.

**Comment by vogelpi** [comment]
Update: After the end of the quiet period, I've rebased this PR on top of `master` (first push) and then applied the patch provided by @jadephilipoom (second push). Adjusted the commit message to reflect that Jade contributed to this as well and added my sign-off line (IIRC this is needed for CI).

@jadephilipoom , do you know whether this needs to be cherry-picked over to the `earlgrey_es` branch? I would guess no.

**Comment by jadephilipoom** [comment]
> @jadephilipoom , do you know whether this needs to be cherry-picked over to the `earlgrey_es` branch? I would guess no.

No, I wouldn't expect it to need cherry-picking.

**Comment by vogelpi** [comment]
Great, thanks for your feedback @jadephilipoom !

**Pull request merged**